### PR TITLE
[WIP] Allow loggers to be restarted on HPC resume

### LINKED ISSUE #6205: MLFlow Logger Makes a New Run When Resuming from hpc Checkpoint
## 🐛 Bug

Currently the `MLFlowLogger` creates a new run when resuming from an hpc checkpoint, e.g., after preemption by slurm and requeuing. Runs are an MLFlow concept that groups things in their UI, so when resuming after requeue, it should really be reusing the run ID. I think this can be patched into the hpc checkpoint using the logger which I believe exposes the run ID. This can also be seen on the `v_num` on the progress bar which changes after preemption (in general that `v_num` probably shouldnt be changing in this case).

### To Reproduce

Use `MLFlowLogger` on a slurm cluster and watch the mlflow UI when preemption happens, there will be a new run created.

### Expected behavior

Runs are grouped neatly on the MLFlow UI

### Environment

* CUDA:
        - GPU:
        - available:         False
        - version:           10.2
* Packages:
        - numpy:             1.20.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.7.1
        - pytorch-lightning: 1.2.0
        - tqdm:              4.57.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.1
        - version:           #1 SMP Thu Jan 21 16:15:07 EST 2021

### PR DESCRIPTION
**Pull request by Queuecumber**
## What does this PR do?

Fixes #6205 

This problem is not restricted to MLFlow as discussed in the issue though, it is a problem with pretty much every logger. 

This PR fixes the problem of loggers not resuming when an experiment running on HPC (i.e., slurm) times out and is requeued. The same experiment is requeued and resumes running, but the loggers were making a new experiment entry on the UI. I fixed this by saving the logger in the HPC checkpoint and then moving the HPC resume call earlier in the trainer code. 

This PR is heavily a WIP and I have marked it as such, there's a few small things missing but mainly I wanted to get the ball rolling on some discussion around it. I'm using this code to train real models and it works fine for me but I would love someone who knows all  the implications of moving the HPC loading code around to have a look at this. I tried to start this discussion on #6205 but I didnt get any replies despite repeatedly asking. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -304,6 +307,9 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
             elif self.trainer.amp_backend == AMPType.APEX:
                 checkpoint['amp_scaling_state'] = amp.state_dict()
 
+            # dump logger
+            checkpoint['logger'] = self.trainer.logger
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:311]: I don't think we should pickle the logger as it will break the checkpoint if their API changes.
Better to save their states instead.

**Review by tchaton — CHANGES_REQUESTED** [review]
We shouldn't pickle the logger but save the logger state such as experiment id, etc..

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 1 above)
Comment (Queuecumber) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:311]: I was originally saving the states but it looks like a number of the loggers have actually taken steps to make sure they are picklable and that they restore correctly (CometLogger being one of them) although I'm not sure what that was done because as far as I can tell there wasn't any saving of loggers happening. 

I like this better because it's simpler and it matches up with the rest of the saving code which seems to prefer pickling objects over saving and restoring states. This way the logic of making object serializable is the responsibility of the person who writes the object (by making it picklable), the internal state of the logger is often not easy to serialize generically, so it's either this or we add a function to the logger API for returning and restoring state that every logger then has to implement. 

Also it wouldn't break the checkpoint because I allow logger loading to fail without disrupting the rest of the resume process. Probably we should also allow logger saving to fail in case a logger isn't picklable.

Comment (Queuecumber) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py]: I took another look at this and decided to go with the state saving API, I'm still working out why some of the tests are failing but please leave your feedback while im doing that. I modeled the API loosely after the callback state saving API

**Comment by Queuecumber** [comment]
I think I resolved the main issue causing failing test in my PR, which was that training checkpoints were being restored even during prediction and evaluation modes. But I am still getting some strange errors in the failing tests, I'll continue to look into them but if anyone knows what might be happening please let me know.

**Comment by Queuecumber** [comment]
Can someone say something about this? I'd like to get it merged and out of the way. @tchaton either resume looking at this or suggest someone that can.

**Comment by Queuecumber** [comment]
@edenlightning can you recommend someone to review this PR please?

**Comment by ashleve** [comment]
@Queuecumber There has been a similar PR here: https://github.com/PyTorchLightning/pytorch-lightning/pull/6361
Currently it's on hold because of a proposition for redesigning Logger class: https://github.com/PyTorchLightning/pytorch-lightning/issues/6606

**Comment by Queuecumber** [comment]
That proposal hasn't had any activity in two months is it worth holding up all logging related PRs over it? 

I agree there's significant overlap between #6361 and this issue so we should convergence on a common solution and make sure the logger state gets written to both types of checkpoints.
